Converter rework

tmadlener opened the pull request:
Changing the converter such that it uses the delphes output TTree that is produced by the `TreeWriter` instead of trying to roll our own version of the `TreeWriter` with a different output EDM. This approach has the advantage, that delphes does most of the heavy lifting for us regarding the finding the generated particles for the reconstructed particles in the output. Additionally, we could also write a rather simple converter that can use already produced delphes output files as inputs and convert them to edm4hep without having to run delphes again. This is not easily possible up to now.

The main changes:
- We now have to patch `ExRootTreeWriter` to be able to manually set a `TTree`. On the other hand we can drop the patches to `Delphes`. In the end I think this small `ExRootTreeWriter` patch could even go to upstream delphes, since it doesn't really affect the default behavior.
- The `DelphesEDM4HepConverter::process` member method now takes a `TTree` as input instead of `Delphes`. This is possible because at the point where the converter now starts processing `Delphes` is already finished and has produced its output.
- It is the responsibility of the `DelphesInputReader` to setup the `ExRootTreeWriter` and the `TTree` to which the output will be written. It also offers access to the `TTree` via a `converterTree` method.

There are still a few things to do, but maybe a discussion about the general design can already be started.

TODO list:
- [x] Update all input readers to the new interface. (Currently only the Pythia reader is done)
- [ ] Some of the delphes output classes do not offer the same information as the `Candidate` class that was used before.
  - [x] `Track`s have no mass. **Setting it to the pion mass**
  - [x] `Tower`s do not expose their kinematics directly. **Using the `Tower::P4` method to get access to the 4 momentum**
  - [ ] `Tower`s have no position information.
- [x] The associations between MC and Reco have to be checked again, as they haven't really been touched yet.
- [x] Ideally there are some tests that we could run in CI that also serve as sort of documentation on how we handle different cases.
- [ ] Make sure documentation is up-to-date.

tmadlener commented:
Cleaned up the internals of the readers and the commit history and this is now in a basic working state again.

Most importantly `Photon`s, `Muon`s and `Electron`s from the Delphes side are now all handled on the same footing. I.e. we take the Delphes candidate and look for a suitable `edm4hep::ReconstructedParticle` to which we set the reference, first by getting a set of candidates via the UniqueID of the generated particle and then from this candidates select one that matches the 4-momentum. This gets rid of all the warnings of the kind

```
**** WARNING: More than one reconstructed particle attached to the genParticle UniqueID 103, which is supposed to be a photon
```

Additionally, the jet constituents that are now deduplicated before they are added to the Jet on the edm4hep side to avoid having the same `ReconstructedParticle` multiple times in the list.

I am still working on some tests that could be run in CI (or maybe on demand?) that also serve as some sort of documentation on how different cases are handled.

vvolkl commented on standalone/src/ExRootTreeWriter.h:
```diff
@@ -0,0 +1,52 @@
+
+  ExRootTreeBranch *NewBranch(const char *name, TClass *cl);
+
+  void SetTree(TTree* t) { fTree = t; }
```
Add a comment to indicate the changes wrt upstream?

tmadlener replied:
Done.

vvolkl reviewed (changes requested):
Thanks Thomas, I'll check more detail in a bit. At one point we should talk again with the delphes developers, because if the intended interface https://cp3.irmp.ucl.ac.be/projects/delphes/wiki/WorkBook/LibraryInterface is not practical, there is something wrong! I'll also try to add some more tests. we can definitely have groups of test that get run at different times, I think one can use labels for that.

clementhelsens commented:
> Cleaned up the internals of the readers and the commit history and this is now in a basic working state again.
> 
> Most importantly `Photon`s, `Muon`s and `Electron`s from the Delphes side are now all handled on the same footing. I.e. we take the Delphes candidate and look for a suitable `edm4hep::ReconstructedParticle` to which we set the reference, first by getting a set of candidates via the UniqueID of the generated particle and then from this candidates select one that matches the 4-momentum. This gets rid of all the warnings of the kind
> 
> ```
> **** WARNING: More than one reconstructed particle attached to the genParticle UniqueID 103, which is supposed to be a photon
> ```
> 
> Additionally, the jet constituents that are now deduplicated before they are added to the Jet on the edm4hep side to avoid having the same `ReconstructedParticle` multiple times in the list.
> 
> I am still working on some tests that could be run in CI (or maybe on demand?) that also serve as some sort of documentation on how different cases are handled.

Hi @tmadlener ,

I'm not quiet sure about the choice made for the matching to generated particles. Indeed, has we have seen that:
 - a reconstructed particle can be matched to several generated particles simply because several generated particles end up in the same calorimeter tower (for example two photons from pi0 decay)
 - a generated particle can be matched to several reconstructed particles. This could occur for example during the particle flow, where you will have a too large momemtum imbalance between your charged component (from the track) and the calorimeter deposit associated to it. Let's imagine you have a 10GeV electron, well measured by the tracker. In your electromagnetic calorimeter, imagine you reconstruct much more (15GeV) and that difference is incompatible with the ECAL resolution, in this case, delphes PFlow will create a 5GeV neutral deposit in the ECAL (so an extra photon). Thus one generated particle is connected to two reconstructed PFlow candidates.

From what I understand from your comment, you have completely removed the possibility to get the extra associations? which is not what I think we want

tmadlener commented:
Hi @clementhelsens,

> From what I understand from your comment, you have completely removed the possibility to get the extra associations?

No, actually all the extra associations are still there. This matching is only done to identify the `Photon`s, `Muon`s and `Electron`s in the list of all reconstructed particles.

These are the steps that are done during the conversion in a bit more detail:

1. All the `GenParticle`s are converted to `edm4hep::MCParticle`. Additionally we store the `UniqueID` of each `GenParticle` to identify them later.
2. All the `Track`s from Delphes are converted into an `edm4hep::ReconstructedParticle` with an attached `edm4hep::Track`. The `UniqueID` of the `GenParticle` associated to the `Track` is used to look up the corresponding `edm4hep::MCParticle` and we fill an `edm4hep::MCRecoParticleAssociation` that connects them on the edm4hep side.
3. All the `Tower`s from Delphes are converted into an `edm4hep::ReconstructedParticle` with an attached `edm4hep::Cluster`. The `UniqueID`s of **all** the `GenParticle` associated to the `Tower` are used to look up the corresponding `edm4hep::MCParticle`s and we fill an `edm4hep::MCRecoParticleAssociation` for all of them.
4. Delphes also additionally delievers a `Photon`, a `Muon` and an `Electron` collection. As far as I understand these are filled from the `Tower` and `Track` collection after running some sort of *identification*. Only for these collections do we do the matching to identify the `edm4hep::ReconstructedParticle` that corresponds to the underlying `Track` or `Tower`. We do this identification and store these elements as `edm4hep::RecoParticleRef`s, because we want to avoid a double counting of these objects, while still being able to access them through a collection with a name.
5. We take the `Jet`s delievered by Delphes and turn them into `edm4hep::ReconstructedParticle`s, where we attach the `Constituents` (from Delphes and converted above in 2 and 3) as `edm4hep::ReconstructedParticle`s. We do not add any association to `edm4hep::MCParticle`s, they can be retrieved by checking the constituents.
6. We fill the `MissingET` as an `edm4hep::ReconstructedParticle` and the `ScalarHT` as an `edm4hep::ParticleID` object.

So the matching is only done for object that Delphes has identified as `Muon`, `Electron` or `Photon`. Additionally, we leave the `edm4hep::ReconstructedParticle` to which they refer untouched, except that we update their mass. In principle it would be possible to run a different identification process over all the `edm4hep::ReconstructedParticle`s to get a (possibly different) set of muons, electrons and photons from that. The `edm4hep::ReconstructedParticle`s have as many relations to an `edm4hep::MCParticle` as the original `Track` or `Tower` did in Delphes.

I have added a point to update the documentation to the TODO list above. Maybe we can use this list here as starting point and see where more clarifications are still necessary.

clementhelsens commented:
Thanks @tmadlener, I think we are good!
I also want to point out here storing additional per event weights, like what could be needed for jet matching within Pythia8. I might open an issue about it

tmadlener commented:
> I also want to point out here storing additional per event weights, like what could be needed for jet matching within Pythia8.

There is already #9, could this fit there? Maybe we can then also discuss how to best store the necessary additional information.

tmadlener commented:
I have now added a first test that uses the Pythia8 based reader and converter to generate some output on the fly that is then compared. The main reason for using Pythia8 for this was basically because we already had a Pythia8 cmd file in the examples folder. I have also copied one delphes card to the tests subfolder, where the only change wrt to how it is shipped from Delphes is is that I fix the RandomSeed and set a maximum number of events.

The basic test setup that I have added should it make pretty easy to add tests using other input formats as well, basically what is left to do is to decide which inputs to use (and where to get them, see #12).

As far as the comparison between the (original) delphes output and the edm4hep output goes,
- I check that the `GenParticle`, `Muon`, `Electron`, `Photon`, `EFlowTrack`, `EFlowPhoton`, `EFlowNeutralHadron` and `Jet` collections have the same contents. This means that for all of these collections I check that all the elements that are present in the delphes output are also present in the edm4hep output, by comparing their kinematics. 
- For the `GenParticle`/`edm4hep::MCParticle` I also check that the mother/daughter relations are the same. For the `EFlowTrack`, `EFlowPhoton`, and `EFlowNeutralHadron` collections that go into the "global" `edm4hep::ReconstructedParticle` collection, I additionally check that they have the same associations to `edm4hep::MCParticle` as they have on the delphes side. 
- For the `Muon`, `Electron`, and `Photon` collections that are stored as `edm4hep::RecoParticleRef` I simply check that the kinematics of the particle they point to matches the ones reported by delphes.
- For the `Jet` collection I check whether they have the same kinematics and also whether their constituents are the same in delphes and edm4hep.

I don't check the details of the `edm4hep::Cluster`s and `edm4hep::Track`s that are attached to the `edm4hep::ReconstructedParticle` since I think that would basically amount to checking getters/setters. For similar reasons I currently also do not check other properties, e.g. the PDG of MCParticles...

I think overall this should test the basic conversion pretty well, and it has already uncovered two unreported bugs:
- The daugther relations are not just the inversions of the mother relations in Delphes. Also in the converter they are now set independently of the mother relations.
- The jet constituents are now properly converted.

I consider this ready for review now and remove the WIP label. In the interest of getting some of the fixes merged, we could also consider adding more test cases and fixing some of the remaining (small) things still in the TODO list for this in another PR.

vvolkl reviewed (approved):
Hi Thomas,

thanks, this is a great improvement. As discussed in the meeting, I'm going to merge this and add my comments / questions as further pull requests.

My very high level comment is that the the converter interface still *should* use Delphes instead of TTree, but it's true that a lot of the conversion code in the Delphes Tree writer then gets duplicated and I'm not sure if you could convert the contents of a delphes ExRootAnalysis File (even though it should be possible, with ExRootTreeReader and ImportArroy).
